packages/ref-core/src/ref_core/metrics.py

```diff
@@ -87,6 +91,79 @@ class TriggerInfo:
     # dataset metadata
 
 
+@frozen(hash=True)
+class DataRequirement:
+    """
+    Definition of the input datasets that a metric requires to run.
+
+    This is used to determine the different metric executions that are required to be performed
+    with the current data catalog.
+
+    The data catalog is filtered according to the `filters` field,
+    then grouped according to the `group_by` field,
+    and then each group is checked that it satisfies the `constraints`.
+    Each such group will be processed as a separate execution of the metric.
+    """
```

Comment:
Maybe we could also just let people provide a grouping function? Most may be happy using pandas' groupby, but some might want to do fancy stuff, and why not?

Reply:
Also, not to derail this, but depending on the people we expect to write DataRequirements, SQL could also be a way to specify filters and groupbys. SQL has a rich language for this stuff, we are probably selecting from a database anyway, and sqlite is faster than anything we can implement using pandas. Much faster, if we include the correct indexes.

Reply:
But GroupOperations might not be easily possible in SQL, e.g. getting the parent.

Reply:
> SQL could also be a way to specify filters and groupbys

There might be optimisations that are used in the backend to load the appropriate  data catalog/groups. Currently this looks like [this](https://github.com/CMIP-REF/cmip-ref/blob/17d5119b2ea96ff61328e2cd52e2da587476f9a5/packages/ref/src/ref/datasets/cmip6.py#L220) with no filters. I've already got a TODO in there suggesting that this might not be the best approach.

The author of these DataRequirements will be the metric package authors (a mix of Scientist and RSEs). My gut says that a micro DSL will be easier to understand and have more guardrails than freeform SQL. Some effort will have to go into sane error messages (column x doesn't exist), logging and I'd probably suggest a CLI entrypoint to perform the solving in a stepwise manner to help verify what is happening. 

This will also require a how-to-guide (notebook) for the docs

It doesn't exclude having a SQLDataRequirement for expert 😉

Reply:
> SQL could also be a way to specify filters and groupbys

There might be optimisations that are used in the backend to load the appropriate data catalog/groups. Currently this looks like [this](https://github.com/CMIP-REF/cmip-ref/blob/17d5119b2ea96ff61328e2cd52e2da587476f9a5/packages/ref/src/ref/datasets/cmip6.py#L220) with no filters. I've already got a TODO in there suggesting that this might not be the best approach.